pkg/payload/precondition/clusterversion/upgradeable_test.go

```diff
@@ -60,50 +22,98 @@ func TestUpgradeableRun(t *testing.T) {
 	}
 
 	tests := []struct {
-		name           string
-		upgradeable    *configv1.ConditionStatus
-		currVersion    string
-		desiredVersion string
-		expected       string
+		name               string
+		upgradeable        *configv1.ConditionStatus
+		currVersion        string
+		desiredVersion     string
+		NonBlockingWarning bool
+		expectedReason     string
+		expected           string
 	}{
 		{
 			name:           "first",
 			desiredVersion: "4.2",
-			expected:       "",
 		},
 		{
 			name:           "move-y, no condition",
 			currVersion:    "4.1",
 			desiredVersion: "4.2",
-			expected:       "",
 		},
 		{
 			name:           "move-y, with true condition",
 			upgradeable:    ptr(configv1.ConditionTrue),
 			currVersion:    "4.1",
 			desiredVersion: "4.2",
-			expected:       "",
 		},
 		{
 			name:           "move-y, with unknown condition",
 			upgradeable:    ptr(configv1.ConditionUnknown),
 			currVersion:    "4.1",
 			desiredVersion: "4.2",
-			expected:       "",
 		},
 		{
-			name:           "move-y, with false condition",
+			name:               "invalid current version",
+			upgradeable:        ptr(configv1.ConditionFalse),
+			currVersion:        "4.1",
+			desiredVersion:     "4.2",
+			NonBlockingWarning: true,
+			expectedReason:     "InvalidCurrentVersion",
+			expected:           "No Major.Minor.Patch elements found",
+		},
+		{
+			name:           "invalid target version",
 			upgradeable:    ptr(configv1.ConditionFalse),
-			currVersion:    "4.1",
+			currVersion:    "4.1.1",
 			desiredVersion: "4.2",
+			expectedReason: "InvalidDesiredVersion",
+			expected:       "No Major.Minor.Patch elements found",
+		},
+		{
+			name:           "move-y, with false condition",
+			upgradeable:    ptr(configv1.ConditionFalse),
+			currVersion:    "4.1.1",
+			desiredVersion: "4.2.1",
 			expected:       "set to False",
 		},
 		{
 			name:           "move-z, with false condition",
 			upgradeable:    ptr(configv1.ConditionFalse),
 			currVersion:    "4.1.3",
 			desiredVersion: "4.1.4",
-			expected:       "",
+		},
+		{
+			name: "empty target",
+		},
+		{
+			name:           "downgrade z",
+			upgradeable:    ptr(configv1.ConditionFalse),
+			currVersion:    "4.16.3",
+			desiredVersion: "4.16.1",
+		},
+		{
+			name:           "downgrade y",
+			upgradeable:    ptr(configv1.ConditionFalse),
+			currVersion:    "4.16.1",
+			desiredVersion: "4.15.3",
+			expected:       "set to False",
```

Comment:
only consuming `len(tc.expected)` is super weird; I hope that wasn't past-me's idea.  But 🤷 , no worries from me if you follow that existing pattern here instead of moving to something more intuitive (like using [`configv1.ConditionStatus`][1]).

[1]: https://pkg.go.dev/github.com/openshift/api/config/v1#ConditionStatus

Reply:
Also, downgrades have [their own guard][1].  It's probably worth adjusting your new logic to allow those, so we don't distract folks attempting a rollback to 4.(y-1) with "and also, here's some concerns about 4.(y+1)..." noise.

[1]: https://github.com/openshift/cluster-version-operator/blob/e546515213c8681ca44c52f178401cd47ad07d11/pkg/payload/precondition/clusterversion/rollback.go#L100

Reply:
This one is  blocked by "upgradeable=false", not downgrading version.
Changed the test name to tell this.

Reply:
Right, I'm saying retargting from `4.16.1` to `4.15.3` has nothing to do with `Upgradeable`, regardless of whether it's `True` or `False`.  _This_ precondition should say "yeah, `Upgradeable` is only concerned about moving forward", and [the rollback precondition][1] will say "I have a problem with this kind of rollback!".

[1]: https://github.com/openshift/cluster-version-operator/blob/e546515213c8681ca44c52f178401cd47ad07d11/pkg/payload/precondition/clusterversion/rollback.go#L100